File: q221.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if(matrix.empty()) return 0;
        int n=matrix.size();
        int m=matrix[0].size();
        
        int ans=0,flag=0;
        vector<vector<int>>res(n,vector<int>(m,0));
        for(int i=0;i<n;i++){
            res[i][0]=matrix[i][0]-'0';
            flag=max(flag,res[i][0]);
        }
        for(int j=0;j<m;j++){
            res[0][j]=matrix[0][j]-'0';
            flag=max(flag,res[0][j]);
        }
        for(int i=1;i<n;i++){
            for(int j=1;j<m;j++){
                if(matrix[i][j]=='0') 
                    res[i][j]=0;
                else 
                    res[i][j]=1+min(res[i-1][j-1],min(res[i-1][j],res[i][j-1]));
                
                if(ans<res[i][j]) ans=res[i][j];
            }
            
        }
  
        return max(ans*ans,flag);
    }
};

File: q221.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        rows=len(matrix)
        cols=len(matrix[0])
        flag,ans=0,0
        res = [[0 for i in range(cols)] for j in range(rows)]
        print(res)
        for i in range(0,rows):
            res[i][0]=int(matrix[i][0])
            flag=max(res[i][0],flag)
            
        for j in range(0,cols):
            res[0][j]=int(matrix[0][j])
            flag=max(res[0][j],flag)
            
        for i in range(1,rows):
            for j in range(1,cols):
                if(matrix[i][j]=='0'):
                    res[i][j]=0
                else:
                    res[i][j]=1+min(res[i-1][j-1],res[i-1][j],res[i][j-1])
                if ans<res[i][j]:
                    ans=res[i][j]
        
        return max(ans*ans,flag)
